xdem/coreg.py

```diff
@@ -1442,3 +1482,426 @@ def _to_matrix_func(self) -> np.ndarray:
+                    "y_off": points[i, 1, 1] - points[i, 1, 0],
+                    "z_off": points[i, 2, 1] - points[i, 2, 0],
+                    "inlier_count": chunk_meta[i]["inlier_count"],
+                    "nmad": chunk_meta[i]["nmad"],
+                    "median": chunk_meta[i]["median"]
+                }
+            )
+
+        stats_df = pd.DataFrame(statistics)
+        stats_df.index.name = "chunk"
+
+        return stats_df
+            
+
+    def subdivide_array(self, shape: tuple[int, ...]) -> np.ndarray:
+        """
+        Return the grid subdivision for a given DEM shape.
+
+        :param shape: The shape of the input DEM.
+        
+        :returns: An array of shape 'shape' with 'self.subdivision' unique indices.
+        """
+        if len(shape) == 3 and shape[0] == 1:  # Account for (1, row, col) shapes
+            shape = (shape[1], shape[2])
+        return xdem.spatial_tools.subdivide_array(shape, count=self.subdivision)
+
+
+    def _apply_func(self, dem: np.ndarray, transform: rio.transform.Affine) -> np.ndarray:
+
+        points = self.to_points()
+
+        bounds, resolution = _transform_to_bounds_and_res(dem.shape, transform)
+        
+        representative_height = np.nanmean(dem)
+        edges_source = np.array([
+            [bounds.left + resolution / 2, bounds.top - resolution / 2, representative_height],
+            [bounds.right - resolution / 2, bounds.top - resolution / 2, representative_height],
+            [bounds.left + resolution / 2, bounds.bottom + resolution / 2, representative_height],
+            [bounds.right - resolution / 2, bounds.bottom + resolution / 2, representative_height]
+        ])
+        edges_dest = self.apply_pts(edges_source)
+        edges = np.dstack((edges_source, edges_dest))
+
+        all_points = np.append(points, edges, axis=0)
+
+        warped_dem = warp_dem(
+            dem=dem,
+            transform=transform,
+            source_coords=all_points[:, :, 0],
+            destination_coords=all_points[:, :, 1],
+            resampling="linear"
+        )
+
+        return warped_dem
+
+
+
+    def _apply_pts_func(self, coords: np.ndarray) -> np.ndarray:
+        """Apply the scaling model to a set of points."""
+        points = self.to_points()
+
+        new_coords = coords.copy()
+
+        for dim in range(0, 3):
+            with warnings.catch_warnings():
+                # ZeroDivisionErrors may happen when the transformation is empty (which is fine)
+                warnings.filterwarnings("ignore", message="ZeroDivisionError")
+                model = scipy.interpolate.Rbf(
+                    points[:, 0, 0],
+                    points[:, 1, 0],
+                    points[:, dim, 1] - points[:, dim, 0],
+                    function="linear",
+                )
+
+            new_coords[:, dim] += model(coords[:, 0], coords[:, 1])
+
+        return new_coords
+
+
+def warp_dem(
```

Comment:
I'm not sure I grasp the addition of `warp_dem` in relation to `geoutils.reproject()` ?

Reply:
Thanks a lot for the feedback, @rhugonnet!

>     * The blocks might need overlapping to avoid introducing edge effects for many type of coregistration methods, such as horizontal shifts.
I'm not sure this is needed, as the blocks are not transformed separately, but a transformation is made using a thin-plate spline approximation. Here's a figure of it that I found on the internets:

![](https://www.researchgate.net/profile/H-Stiehl/publication/11904171/figure/fig5/AS:668834281320460@1536473927530/a-d-Performance-of-approximating-thin-plate-splines-visualized-by-deforming-a-regular_W640.jpg)

Basically, the transformation matrix is described as an X/Y/Z shift near-center of every block, so for e.g. a subdivision of 64, we have 64 points telling an algorithm "these points here should rather be X px to the right, Y px north, etc.). Then I use `xdem.coreg.warp_dem` to do a "puppet warp" (as it's called in Photoshop) on the DEM, along with a vertical correction as well (which you can see in the upper left plot in the figure). This means that there will be no seams between the blocks, as the warping is performed on the whole raster at the same time. 

Does that make sense?

>     * There is existing, efficient tools made for parallel processing/lazy computing which have been designed to improve both the speed and the memory usage of such methods, specifically for numpy arrays in **Dask**. We'll need to think about switching to those soon, otherwise our methods will become obsolete quite rapidly!
Interesting! Do we already have Dask as a dependency? `concurrent.futures` works pretty well together with some copying of only the chunks' interiors (I never copy the whole raster in a thread), so I would say that this is pretty solid. But if there's a much better way, we should of course embrace it. I just wonder, is Dask a heavy dependency and should it be required? What do you think, @adehecq?

> Regarding the figure example you present: what is the exact time of acquisition for those DEMs? did you consider that the residuals of the Nuth and Kääb could be due to different amounts of snow-height in the surrounding of the glaciers?
> Fig. S5 here: https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-021-03436-z/MediaObjects/41586_2021_3436_MOESM1_ESM.pdf

I have used the images myself for my MSc thesis, and the snow cover is almost identical. There might be a tiny bit more in 2009, but the potential difference would be on decimeter level. Here, we see differences of +2 m to -4 m, which is way too much!